File: sdk/kotlin-enhanced/src/jvmMain/kotlin/com/flagent/enhanced/manager/FlagentManager.kt
package com.flagent.enhanced.manager

import com.flagent.client.apis.EvaluationApi
import com.flagent.client.models.EvalContext
import io.ktor.client.statement.bodyAsText
import com.flagent.client.models.EvalResult
import com.flagent.client.models.EvaluationBatchRequest
import com.flagent.client.models.EvaluationEntity
import com.flagent.enhanced.cache.CacheKey
import com.flagent.enhanced.cache.EvaluationCache
import com.flagent.enhanced.cache.InMemoryEvaluationCache
import com.flagent.enhanced.config.FlagentConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

class FlagentManager(
    private val evaluationApi: EvaluationApi,
    private val config: FlagentConfig = FlagentConfig()
) {
    private val cache: EvaluationCache? = if (config.enableCache) {
        InMemoryEvaluationCache(config.cacheTtlMs)
    } else null

    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    init {
        if (cache != null) {
            scope.launch {
                while (true) {
                    delay(config.cacheTtlMs)
                    cache.evictExpired()
                }
            }
        }
    }

    suspend fun evaluate(
        flagKey: String? = null,
        flagID: Long? = null,
        entityID: String? = null,
        entityType: String? = null,
        entityContext: Map<String, Any>? = null,
        enableDebug: Boolean = false
    ): EvalResult = withContext(Dispatchers.Default) {
        val cacheKey = CacheKey(flagKey, flagID, entityID, entityType)
        val cached = cache?.get(cacheKey)
        if (cached != null) return@withContext cached
        val jsonEntityContext: JsonObject? = entityContext?.let { ctx ->
            buildJsonObject {
                ctx.forEach { (key, value) ->
                    put(key, when (value) {
                        is String -> JsonPrimitive(value)
                        is Number -> JsonPrimitive(value)
                        is Boolean -> JsonPrimitive(value)
                        is kotlinx.serialization.json.JsonElement -> value
                        else -> JsonPrimitive(value.toString())
                    })
                }
            }
        }
        val evalContext = EvalContext(
            flagKey = flagKey,
            flagID = flagID,
            entityID = entityID,
            entityType = entityType,
            entityContext = jsonEntityContext,
            enableDebug = enableDebug
        )
        val response = evaluationApi.postEvaluation(evalContext)
        if (!response.success) {
            val body = response.response.bodyAsText()
            throw IllegalStateException("Evaluation failed: ${response.status} - $body")
        }
        val result = response.body()
        cache?.put(cacheKey, result)
        result
    }

    suspend fun evaluateBatch(
        flagKeys: List<String>? = null,
        flagIDs: List<Int>? = null,
        entities: List<EvaluationEntity>,
        enableDebug: Boolean = false
    ): List<EvalResult> = withContext(Dispatchers.Default) {
        val request = EvaluationBatchRequest(
            entities = entities,
            flagKeys = flagKeys,
            flagIDs = flagIDs,
            enableDebug = enableDebug
        )
        val response = evaluationApi.postEvaluationBatch(request)
        if (!response.success) {
            val body = response.response.bodyAsText()
            throw IllegalStateException("Batch evaluation failed: ${response.status} - $body")
        }
        response.body().evaluationResults
    }

    suspend fun clearCache() {
        cache?.clear()
    }

    suspend fun evictExpired() {
        cache?.evictExpired()
    }
}

File: sdk/flutter-enhanced/lib/src/manager/flagent_manager.dart
import 'dart:async';

import 'package:built_collection/built_collection.dart';
import 'package:built_value/json_object.dart';
import 'package:dio/dio.dart';
import 'package:flagent_client/flagent_client.dart';

import '../cache/evaluation_cache.dart';
import '../config/flagent_config.dart';

BuiltMap<String, JsonObject?>? _toEntityContext(Map<String, dynamic>? ctx) {
  if (ctx == null || ctx.isEmpty) return null;
  final builder = MapBuilder<String, JsonObject?>();
  for (final e in ctx.entries) {
    if (e.value != null) {
      builder[e.key] = JsonObject(_toJsonValue(e.value));
    }
  }
  return builder.build();
}

Object _toJsonValue(dynamic v) {
  if (v is String || v is num || v is bool) return v;
  if (v is List) return v.map(_toJsonValue).toList();
  if (v is Map) return v.map((k, v) => MapEntry(k.toString(), _toJsonValue(v)));
  return v.toString();
}

/// Build EvaluationEntity from simple parameters.
EvaluationEntity buildEvaluationEntity({
  required String entityID,
  String? entityType,
  Map<String, dynamic>? entityContext,
}) {
  return EvaluationEntity((b) {
    b
      ..entityID = entityID
      ..entityType = entityType;
    final ctx = _toEntityContext(entityContext);
    if (ctx != null) b.entityContext.replace(ctx);
  });
}

/// Enhanced Flagent Manager with caching and convenient API.
///
/// Provides high-level API for evaluating feature flags with automatic caching.
class FlagentManager {
  final EvaluationApi _evaluationApi;
  final FlagentConfig _config;
  final EvaluationCache? _cache;
  Timer? _cleanupTimer;

  FlagentManager(
    String basePath, {
    Dio? dio,
    FlagentConfig config = defaultFlagentConfig,
  })  : _config = config,
        _evaluationApi = FlagentClient(
          basePathOverride: basePath,
          dio: dio,
        ).getEvaluationApi(),
        _cache = config.enableCache
            ? InMemoryEvaluationCache(config.cacheTtlMs)
            : null {
    if (_cache != null) {
      _cleanupTimer = Timer.periodic(
        Duration(milliseconds: config.cacheTtlMs),
        (_) => _cache!.evictExpired(),
      );
    }
  }

  /// Evaluate a flag for a given entity context.
  Future<EvalResult> evaluate({
    String? flagKey,
    int? flagID,
    String? entityID,
    String? entityType,
    Map<String, dynamic>? entityContext,
    bool enableDebug = false,
  }) async {
    final cacheKey = CacheKey(
      flagKey: flagKey,
      flagID: flagID,
      entityID: entityID,
      entityType: entityType,
    );

    if (_cache != null) {
      final cached = await _cache!.get(cacheKey);
      if (cached != null) return cached;
    }

    final evalContext = EvalContext((b) {
      b
        ..flagKey = flagKey
        ..flagID = flagID
        ..entityID = entityID
        ..entityType = entityType
        ..enableDebug = enableDebug;
      final ctx = _toEntityContext(entityContext);
      if (ctx != null) b.entityContext.replace(ctx);
    });

    final response = await _evaluationApi.postEvaluation(
      evalContext: evalContext,
    );

    if (response.data == null) {
      throw Exception('Evaluation failed: ${response.statusCode}');
    }

    final result = response.data!;
    if (_cache != null) {
      await _cache!.put(cacheKey, result);
    }
    return result;
  }

  /// Batch evaluate flags for multiple entities.
  Future<List<EvalResult>> evaluateBatch({
    List<String>? flagKeys,
    List<int>? flagIDs,
    required List<EvaluationEntity> entities,
    bool enableDebug = false,
  }) async {
    final request = EvaluationBatchRequest((b) => b
      ..entities = ListBuilder(entities)
      ..flagKeys = flagKeys != null ? ListBuilder(flagKeys) : null
      ..flagIDs = flagIDs != null ? ListBuilder(flagIDs) : null
      ..enableDebug = enableDebug);

    final response = await _evaluationApi.postEvaluationBatch(
      evaluationBatchRequest: request,
    );

    if (response.data == null) {
      throw Exception('Batch evaluation failed: ${response.statusCode}');
    }

    return response.data!.evaluationResults.toList();
  }

  /// Clear all cached evaluation results.
  Future<void> clearCache() async {
    await _cache?.clear();
  }

  /// Evict expired entries from cache.
  Future<void> evictExpired() async {
    await _cache?.evictExpired();
  }

  /// Destroy the manager and clean up resources.
  void destroy() {
    _cleanupTimer?.cancel();
    _cleanupTimer = null;
  }
}
